streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -272,15 +280,22 @@ public synchronized void onChange(final StreamThread thread,
             }
             if (newState == StreamThread.State.PARTITIONS_REVOKED ||
                 newState == StreamThread.State.ASSIGNING_PARTITIONS) {
-                setState(State.REBALANCING);
+                setStateWhenNotShuttingDownOrNotRunning(State.REBALANCING);
             } else if (newState == StreamThread.State.RUNNING) {
+                // one thread is running, check others
                 for (final StreamThread.State state : threadState.values()) {
                     if (state != StreamThread.State.RUNNING) {
                         return;
                     }
                 }
-                setState(State.RUNNING);
+                setStateWhenNotShuttingDownOrNotRunning(State.RUNNING);
```

Comment:
I don't think this is related to this line, but the scenario you mention can happen. The transition of the instance would be RUNNING->REBALANCING->RUNNING. It's ok to have 1 thread to DEAD, as long as another thread is up.